### PR TITLE
Fusion: initial PDF

### PR DESCRIPTION
**Pull request by JeffGainsNGames**

### CODE REVIEW

**Review by Miepee — CHANGES_REQUESTED** [review]
please rebase the pdf to have/use the new functions there were named after before I give this a proper review. Look at any of the existing pdfs if you need a reference.

**Comment by Miepee** [comment]
Also please add tests.

File: randovania/games/fusion/exporter/patch_data_factory.py [hunk 1]
```diff
@@ -1,13 +1,253 @@
...
+            state
+            for defi, state in self.configuration.standard_pickup_configuration.pickups_state.items()
+            if defi.name == "Power Bomb Data"
+        )
+        starting_dict["PowerBombs"] = pb_launcher.included_ammo[0]
+
+        for item in self.patches.starting_equipment:
+            if "Metroid" in item.name:
+                continue
+            pickup_def = next(
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/fusion/exporter/patch_data_factory.py:89]: if I did this in the initial pdf the reason why is probably my usual "it worked", if you did this I have no idea either. Can this be pushed off to a future PR?

File: randovania/games/fusion/exporter/patch_data_factory.py [hunk 2]
```diff
@@ -1,13 +1,253 @@
...
+            "RequiredMetroidCount": self.configuration.artifacts.required_artifacts,
+            "TankIncrements": self._create_tank_increments(),
+            "DoorLocks": self._create_door_locks(),
+            "SkipDoorTransitions": self.configuration.instant_transitions,
+            "Palettes": self._create_palette(),
+            "NavigationText": self._create_nav_text(),
+        }
+        import json
+
+        foo = json.dumps(mars_data)
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/fusion/exporter/patch_data_factory.py:251]: not exactly sure what that means, can probably remove these lines too right since this was just for testing?

[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/fusion/exporter/patch_data_factory.py:251]: Inlining = replace all calls with a certain thing, by its contents. Content can either be an assignment if you're inlining variables, or a function body, if you're inlining functions.
TLDR; print(json.dumps(mars_data))

File: randovania/games/fusion/exporter/patch_data_factory.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/fusion/exporter/patch_data_factory.py:89]: I don't see a reason why we should push this to a future PR.
If you have no idea why we did this, and I have no idea why we did this, then it sounds like a perfect opportunity to rewrite that.

File: randovania/games/fusion/exporter/patch_data_factory.py [hunk 3]
```diff
@@ -1,13 +1,249 @@
...
+        )
+
+    def create_visual_nothing(self) -> PickupEntry:
+        """The model of this pickup replaces the model of all pickups when PickupModelDataSource is ETM"""
+        return pickup_creator.create_visual_nothing(self.game_enum(), "Empty")
+
     def create_game_specific_data(self) -> dict:
+        pickup_list = self.export_pickup_list()
+
+        mars_data = {
```
Comment (Miepee) [inline randovania/games/fusion/exporter/patch_data_factory.py:232]: Todo: add credits

File: randovania/games/fusion/exporter/patch_data_factory.py [hunk 4]
```diff
@@ -1,13 +1,249 @@
...
+            model_name="Empty",
+        )
+
+    def create_visual_nothing(self) -> PickupEntry:
+        """The model of this pickup replaces the model of all pickups when PickupModelDataSource is ETM"""
+        return pickup_creator.create_visual_nothing(self.game_enum(), "Empty")
+
     def create_game_specific_data(self) -> dict:
+        pickup_list = self.export_pickup_list()
+
```
Comment (Miepee) [inline randovania/games/fusion/exporter/patch_data_factory.py:231]: We have options for these but youre not passing them through:
- stereo, mute music, mute sfx

File: randovania/games/fusion/exporter/patch_data_factory.py (hunk 3 above)
Comment (Miepee) [inline randovania/games/fusion/exporter/patch_data_factory.py:232]: Currently mars supports things that we dont. Writing them down here so that we have them collected. Should probably discuss elsewhere which options we want to provide / always set:
- option to disable demos on title screen
- Make door transitions instant instead of fast
- vanilla / "mars / zm missile limit
- unexplored map

File: randovania/games/fusion/exporter/patch_data_factory.py [hunk 5]
```diff
@@ -1,13 +1,249 @@
...
+        for item in self.patches.starting_equipment:
+            if "Metroid" in item.name:
+                continue
+            category = item.progression[0][0].extra["StartingItemCategory"]
+            # Special Case for E-Tanks
+            if category == "Energy":
+                starting_dict[category] += self.configuration.energy_per_tank
+                continue
+            elif category in ["Abilities", "SecurityLevels"]:
+                starting_dict[category].append(item.progression[0][0].extra["item"])
```
[the remark this answers is not in the document]
Comment (JeffGainsNGames) [reply randovania/games/fusion/exporter/patch_data_factory.py:97]: when I remove this it tries to put missiles category in here, but that is already handled above.

File: randovania/games/fusion/exporter/patch_data_factory.py (hunk 3 above)
Comment (JeffGainsNGames) [reply randovania/games/fusion/exporter/patch_data_factory.py:232]: skip door transitions is in

File: randovania/games/fusion/exporter/patch_data_factory.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Miepee) [reply randovania/games/fusion/exporter/patch_data_factory.py:97]: then do this instead:
```
# Special case for ammo and etanks.
if category in ["Missiles", "PowerBombs"]:
  continue
elif category == "Energy":
  ...

# Normal case
```

**Review by Miepee — APPROVED** [review]
